File: LeetCode/easy/Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public: //递归也可以
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL) return head;
        ListNode* temp = head;
        head = temp->next;
        temp->next = head->next;
        head->next = temp;
        while (temp->next != NULL && temp->next->next != NULL) {
            ListNode* temp1 = temp->next;
            temp->next = temp1->next;
            temp1->next = temp->next->next;
            temp->next->next = temp1;
            temp = temp->next->next;
        }
        return head;
    }
};

File: Lintcode/451_Swap Nodes in Pairs/Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

public class Solution {
	/*
	 * @param head: a ListNode
	 * 
	 * @return: a ListNode
	 */
	public ListNode swapPairs(ListNode head) {
		if (head == null || head.next == null)
			return head;
		ListNode t = head.next;
		head.next = t.next;
		t.next = head;
		head = t;

		ListNode temp = head.next;
		ListNode temp2 = temp.next;
		while (temp2 != null && temp2.next != null) {
			temp.next = temp2.next;
			temp2.next = temp2.next.next;
			temp.next.next = temp2;
			temp = temp2;
			temp2 = temp2.next;
		}
		return head;
	}

}